src/passes/SafeHeap.cpp

```diff
@@ -0,0 +1,316 @@
+    func->name = getLoadName(curr);
+    func->params.push_back(i32); // pointer
+    func->params.push_back(i32); // offset
+    func->vars.push_back(i32); // pointer + offset
+    func->result = curr->type;
+    Builder builder(*module);
+    auto* block = builder.makeBlock();
+    block->list.push_back(
+      builder.makeSetLocal(
+        2,
+        builder.makeBinary(
+          AddInt32,
+          builder.makeGetLocal(0, i32),
+          builder.makeGetLocal(1, i32)
+        )
+      )
+    );
+    // check for reading past valid memory: if pointer + offset + bytes
+    block->list.push_back(
+      makeBoundsCheck(curr->type, builder, 2)
+    );
+    // check proper alignment
+    if (curr->align > 1) {
+      block->list.push_back(
+        makeAlignCheck(curr->align, builder, 2)
+      );
+    }
+    // do the load
+    auto* load = module->allocator.alloc<Load>();
+    *load = *curr; // basically the same as the template we are given!
+    load->ptr = builder.makeGetLocal(2, i32);
+    block->list.push_back(load);
+    block->finalize(curr->type);
+    func->body = block;
+    module->addFunction(func);
+  }
+
+  void addStoreFunc(Store* curr, Module* module) {
+    auto* func = new Function;
+    func->name = getStoreName(curr);
+    func->params.push_back(i32); // pointer
+    func->params.push_back(i32); // offset
+    func->params.push_back(curr->valueType); // value
+    func->vars.push_back(i32); // pointer + offset
+    func->result = none;
+    Builder builder(*module);
+    auto* block = builder.makeBlock();
+    block->list.push_back(
+      builder.makeSetLocal(
+        3,
+        builder.makeBinary(
+          AddInt32,
+          builder.makeGetLocal(0, i32),
+          builder.makeGetLocal(1, i32)
+        )
+      )
+    );
+    // check for reading past valid memory: if pointer + offset + bytes
+    block->list.push_back(
+      makeBoundsCheck(curr->valueType, builder, 3)
+    );
+    // check proper alignment
+    if (curr->align > 1) {
+      block->list.push_back(
+        makeAlignCheck(curr->align, builder, 3)
+      );
+    }
+    // do the store
+    auto* store = module->allocator.alloc<Store>();
+    *store = *curr; // basically the same as the template we are given!
+    store->ptr = builder.makeGetLocal(3, i32);
+    store->value = builder.makeGetLocal(2, curr->valueType);
+    block->list.push_back(store);
+    block->finalize(none);
+    func->body = block;
+    module->addFunction(func);
+  }
+
+  Expression* makeAlignCheck(Address align, Builder& builder, Index local) {
+    return builder.makeIf(
+      builder.makeBinary(
+        AndInt32,
+        builder.makeGetLocal(local, i32),
+        builder.makeConst(Literal(int32_t(align - 1)))
+      ),
+      builder.makeCallImport(alignfault, {}, none)
+    );
+  }
+
+  Expression* makeBoundsCheck(WasmType type, Builder& builder, Index local) {
+    return builder.makeIf(
+      builder.makeBinary(
+        OrInt32,
+        builder.makeBinary(
```

Comment:
Seems like this would be even better if we could check for a value less than emscripten's global base. Could we bake that value into this generated code, since the global base is set at compile time anyway?

Reply:
We don't have that information here, but it's an interesting idea. Not easy though since in asm2wasm we do use lower locations for some special things (like mapped globals, before we started to use wasm globals, and I'm not sure there are no remnants of that; another thing is the special tempDoublePtr stuff that emscripten uses to emulate some things). Can investigate in the future.